File: app-components/src/test/java/org/synsystems/onlypass/components/cryptography/credentialhardening/TestPbkdf2WithHmacSha256CredentialHardener.java
package org.synsystems.onlypass.components.cryptography.credentialhardening;

import org.junit.Before;
import org.junit.Test;
import org.synsystems.onlypass.components.cryptography.credentials.CleartextPassword;

import java.security.NoSuchAlgorithmException;

import static java.util.concurrent.TimeUnit.MILLISECONDS;
import static org.mockito.Mockito.mock;

@SuppressWarnings("ConstantConditions")
public class TestPbkdf2WithHmacSha256CredentialHardener {
  private Pbkdf2WithHmacSha256CredentialHardener credentialHardener;

  @Before
  public void setup() throws NoSuchAlgorithmException {
    credentialHardener = new Pbkdf2WithHmacSha256CredentialHardener();
  }

  @Test(expected = RuntimeException.class)
  public void testDeriveSecureCredential_nullInsecureCredential() {
    credentialHardener.hardenCredential(null, mock(Pbkdf2WithHmacSha256Configuration.class));
  }

  @Test(expected = RuntimeException.class)
  public void testDeriveSecureCredential_nullHashingSalt() {
    credentialHardener.hardenCredential(mock(CleartextPassword.class), null);
  }

  @Test
  public void testDeriveSecureCredential_allValidValues() {
    final Pbkdf2WithHmacSha256Configuration parameters = Pbkdf2WithHmacSha256Configuration
        .builder()
        .setSalt(new byte[]{1})
        .setDerivedKeyBitlength(256)
        .setIterationCount(1000)
        .build();

    credentialHardener
        .hardenCredential(CleartextPassword.create("password"), parameters)
        .test()
        .awaitDone(200, MILLISECONDS)
        .assertNoErrors()
        .assertValue(value -> value != null)
        .assertComplete();
  }
}

File: app-components/src/test/java/org/synsystems/onlypass/components/cryptography/credentialhardening/TestPbkdf2WithHmacSha256CredentialHardener.kt
package org.synsystems.onlypass.components.cryptography.credentialhardening

import org.junit.Before
import org.junit.Test
import org.synsystems.onlypass.components.cryptography.credentials.CleartextPassword
import java.security.NoSuchAlgorithmException
import java.util.concurrent.TimeUnit.MILLISECONDS

class TestPbkdf2WithHmacSha256CredentialHardener {
  private lateinit var credentialHardener: Pbkdf2WithHmacSha256CredentialHardener

  @Before
  @Throws(NoSuchAlgorithmException::class)
  fun setup() {
    credentialHardener = Pbkdf2WithHmacSha256CredentialHardener()
  }

  @Test
  fun testDeriveSecureCredential_allValidValues() {
    val parameters = Pbkdf2WithHmacSha256Configuration(arrayOf(1), 256, 1000)
    val originalPassword = CleartextPassword("password")

    credentialHardener
        .hardenCredential(originalPassword, parameters)
        .test()
        .awaitDone(200, MILLISECONDS)
        .assertNoErrors()
        .assertComplete()
  }
}